core/Network/TLS/Handshake/Server.hs

```diff
@@ -315,9 +315,24 @@ doHandshake sparams mcred ctx chosenVersion usedCipher usedCompression clientSes
             usingHState ctx $ setDHPrivate priv
             return serverParams
 
+        decideHash sigAlg = do
+            usedVersion <- usingState_ ctx getVersion
+            case usedVersion of
+              TLS12 -> do
+                  let sHashSigs = supportedHashSignatures $ ctxSupported ctx
+                      cHashSigs = case extensionLookup extensionID_SignatureAlgorithms exts >>= extensionDecode False of
+                          Just (SignatureAlgorithms hss) -> hss
+                          Nothing                        -> []
+                      hashSigs = cHashSigs `intersect` sHashSigs
+                  case filter ((==) sigAlg . snd) hashSigs of
+                      []  -> error ("no hash signature for " ++ show sigAlg)
+                      x:_ -> return $ Just (fst x)
```

Comment:
Yes, the values are in the client's (descending) preference order, servers also have a (descending) preference order, and many servers will want to choose to ignore the client's order and go with their own. When we ignore the client's order for ciphers and use our own, we should also ignore the client's signature algorithm order, and use our (server's) own.

As for inconsistency, it is possible to choose a ciphersuite such that the server's hash/signature pairs compatible with that ciphersuite don't match any of the client's. In such cases a different ciphersuite should have been chosen.

The cipher selection process becomes more complicated:

   * Choose a certificate whose signature algorithm is compatible with the most preferred shared hash/signature algorithm pair (ordered by client or server preference as appropriate).
  * Only then, if the chosen ciphersuite uses DHE or ECDHE, knowing that the ciphersuite is compatible with some client hash/sig pair, choose the corresponding hash/sig pair.